File: app/analytics.py
"""
Deterministic Business Intelligence Analytics Engine using pandas
Guarantees 100% calculation accuracy with zero AI arithmetic hallucination.
"""

from pathlib import Path
from typing import Dict, Any, Optional
import pandas as pd
from app.normalize import normalize_deals, normalize_work_orders
from app.monday import fetch_all_board_items
from app.config import DEALS_BOARD_ID, WORK_ORDERS_BOARD_ID

# In-memory session cache
_CACHE = {
    "deals_df": pd.DataFrame(),
    "work_orders_df": pd.DataFrame(),
    "deal_warnings": [],
    "wo_warnings": [],
    "source": "None",
    "last_sync": None,
}


def load_local_excel() -> Dict[str, Any]:
    """Fallback: Load and normalize datasets from local Excel files."""
    base_dir = Path(__file__).resolve().parent.parent.parent
    
    deal_file = base_dir / "Deal funnel Data.xlsx"
    wo_file = base_dir / "Work_Order_Tracker Data.xlsx"
    
    # Read sheets
    deals_raw = pd.read_excel(deal_file, sheet_name=0)
    wo_raw = pd.read_excel(wo_file, sheet_name=0, header=1)

    clean_deals, d_warn = normalize_deals(deals_raw)
    clean_wos, wo_warn = normalize_work_orders(wo_raw)

    _CACHE["deals_df"] = clean_deals
    _CACHE["work_orders_df"] = clean_wos
    _CACHE["deal_warnings"] = d_warn
    _CACHE["wo_warnings"] = wo_warn
    _CACHE["source"] = "Excel (Fast Boot / Fallback)"
    _CACHE["last_sync"] = pd.Timestamp.now().isoformat()

    return _CACHE


def refresh_data() -> Dict[str, Any]:
    """Fetch live from monday.com API or fall back to Excel."""
    if not DEALS_BOARD_ID or not WORK_ORDERS_BOARD_ID:
        return load_local_excel()

    try:
        raw_deals = fetch_all_board_items(DEALS_BOARD_ID)
        raw_wos = fetch_all_board_items(WORK_ORDERS_BOARD_ID)

        df_deals = pd.DataFrame(raw_deals)
        df_wos = pd.DataFrame(raw_wos)

        clean_deals, d_warn = normalize_deals(df_deals)
        clean_wos, wo_warn = normalize_work_orders(df_wos)

        _CACHE["deals_df"] = clean_deals
        _CACHE["work_orders_df"] = clean_wos
        _CACHE["deal_warnings"] = d_warn
        _CACHE["wo_warnings"] = wo_warn
        _CACHE["source"] = "monday.com (Live GraphQL API)"
        _CACHE["last_sync"] = pd.Timestamp.now().isoformat()

        return _CACHE
    except Exception as e:
        print(f"[Analytics] monday.com fetch warning ({e}). Falling back to local Excel.")
        return load_local_excel()


def get_data() -> Dict[str, Any]:
    """Ensure data is loaded in memory."""
    if _CACHE["deals_df"].empty or _CACHE["work_orders_df"].empty:
        load_local_excel()
    return _CACHE


# ══════════════════════════════════════════════════════════════════════════
# ANALYTICS FUNCTIONS (PANDAS CALCULATIONS)
# ══════════════════════════════════════════════════════════════════════════

def query_deals(
    sector: Optional[str] = None,
    deal_status: Optional[str] = None,
    deal_stage: Optional[str] = None,
    owner_code: Optional[str] = None,
    group_by: Optional[str] = None,
    limit: int = 10,
) -> Dict[str, Any]:
    """Filter and aggregate sales pipeline deals with pandas."""
    data = get_data()
    df: pd.DataFrame = data["deals_df"].copy()

    if sector and sector.lower() != "all":
        df = df[df["Sector"].str.contains(sector, case=False, na=False)]
    if deal_status and deal_status.lower() != "all":
        df = df[df["Deal Status"].str.lower() == deal_status.lower()]
    if deal_stage and deal_stage.lower() != "all":
        df = df[df["Deal Stage"].str.contains(deal_stage, case=False, na=False)]
    if owner_code and owner_code.lower() != "all":
        df = df[df["Owner Code"] == owner_code]

    total_deals = len(df)
    total_val = float(df["Deal Value"].sum())
    won_df = df[df["isWon"]]
    open_df = df[df["isOpen"]]
    lost_df = df[df["isLost"]]
    win_rate = round((len(won_df) / total_deals * 100)) if total_deals > 0 else 0

    # Sector breakdown
    sector_agg = (
        df.groupby("Sector")["Deal Value"]
        .agg(["count", "sum"])
        .reset_index()
        .sort_values(by="sum", ascending=False)
    )
    sector_breakdown = [
        {"sector": r["Sector"], "count": int(r["count"]), "value": round(float(r["sum"]))}
        for _, r in sector_agg.iterrows()
    ]

    # Stage breakdown
    stage_agg = (
        df.groupby("Deal Stage")["Deal Value"]
        .agg(["count", "sum"])
        .reset_index()
        .sort_values(by="sum", ascending=False)
    )
    stage_breakdown = [
        {"stage": r["Deal Stage"], "count": int(r["count"]), "value": round(float(r["sum"]))}
        for _, r in stage_agg.iterrows()
    ]

    # Top deals
    top_deals = (
        df.sort_values(by="Deal Value", ascending=False)
        .head(limit)[["Deal Name", "Client Code", "Sector", "Deal Stage", "Deal Status", "Deal Value", "Closure Probability"]]
        .to_dict(orient="records")
    )
    for d in top_deals:
        d["Deal Value"] = round(float(d["Deal Value"]))

    return {
        "summary": {
            "totalDeals": total_deals,
            "totalPipelineValue": round(total_val),
            "wonDealsCount": len(won_df),
            "wonValue": round(float(won_df["Deal Value"].sum())),
            "openDealsCount": len(open_df),
            "openValue": round(float(open_df["Deal Value"].sum())),
            "lostDealsCount": len(lost_df),
            "winRate": win_rate,
        },
        "stageBreakdown": stage_breakdown,
        "sectorBreakdown": sector_breakdown,
        "topDeals": top_deals,
        "dataWarnings": data["deal_warnings"][:3],
    }


def query_work_orders(
    sector: Optional[str] = None,
    execution_status: Optional[str] = None,
    invoice_status: Optional[str] = None,
    group_by: Optional[str] = None,
    limit: int = 10,
) -> Dict[str, Any]:
    """Filter and aggregate operational work orders and billing metrics."""
    data = get_data()
    df: pd.DataFrame = data["work_orders_df"].copy()

    if sector and sector.lower() != "all":
        df = df[df["Sector"].str.contains(sector, case=False, na=False)]
    if execution_status and execution_status.lower() != "all":
        df = df[df["Execution Status"].str.contains(execution_status, case=False, na=False)]
    if invoice_status and invoice_status.lower() != "all":
        df = df[df["Invoice Status"].str.contains(invoice_status, case=False, na=False)]

    total_contract = float(df["Amount Incl GST"].sum())
    total_billed = float(df["Billed Incl GST"].sum())
    total_collected = float(df["Collected Amount"].sum())
    total_receivable = float(df["Amount Receivable"].sum())

    completed = len(df[df["isCompleted"]])
    ongoing = len(df[df["isOngoing"]])
    stuck = len(df[df["isStuck"]])

    # Sector breakdown with pandas
    sec_agg = (
        df.groupby("Sector")
        .agg({
            "Amount Incl GST": ["count", "sum"],
            "Billed Incl GST": "sum",
            "Collected Amount": "sum",
            "Amount Receivable": "sum",
        })
        .reset_index()
    )
    sec_agg.columns = ["Sector", "count", "contract", "billed", "collected", "receivable"]
    sector_breakdown = [
        {
            "sector": r["Sector"],
            "count": int(r["count"]),
            "contractValue": round(float(r["contract"])),
            "billedValue": round(float(r["billed"])),
            "collectedValue": round(float(r["collected"])),
            "receivableValue": round(float(r["receivable"])),
            "collectionRate": round(float(r["collected"]) / float(r["billed"]) * 100) if float(r["billed"]) > 0 else 0,
        }
        for _, r in sec_agg.sort_values(by="contract", ascending=False).iterrows()
    ]

    # Top outstanding receivables
    top_receivables = (
        df[df["Amount Receivable"] > 0]
        .sort_values(by="Amount Receivable", ascending=False)
        .head(limit)[["Deal Name", "Customer Code", "Sector", "Execution Status", "Amount Incl GST", "Billed Incl GST", "Collected Amount", "Amount Receivable"]]
        .to_dict(orient="records")
    )
    for w in top_receivables:
        for k in ["Amount Incl GST", "Billed Incl GST", "Collected Amount", "Amount Receivable"]:
            w[k] = round(float(w[k]))

    return {
        "summary": {
            "totalWorkOrders": len(df),
            "totalContractValue": round(total_contract),
            "totalBilled": round(total_billed),
            "totalCollected": round(total_collected),
            "totalReceivable": round(total_receivable),
            "completedCount": completed,
            "ongoingCount": ongoing,
            "stuckCount": stuck,
            "billingRate": round(total_billed / total_contract * 100) if total_contract > 0 else 0,
            "collectionRate": round(total_collected / total_billed * 100) if total_billed > 0 else 0,
        },
        "sectorBreakdown": sector_breakdown,
        "topReceivables": top_receivables,
        "dataWarnings": data["wo_warnings"][:3],
    }


def cross_board_summary(sector: Optional[str] = None) -> Dict[str, Any]:
    """Correlate sales pipeline against operational billing performance."""
    deals = query_deals(sector=sector)
    wos = query_work_orders(sector=sector)

    return {
        "sector": sector or "All Sectors",
        "pipeline": deals["summary"],
        "operations": wos["summary"],
        "sectorPipeline": deals["sectorBreakdown"],
        "sectorOperations": wos["sectorBreakdown"],
        "note": "Cross-board correlation performed by industry sector vertical.",
    }


def get_kpis() -> Dict[str, Any]:
    """Return top-level telemetry metrics."""
    data = get_data()
    d_df = data["deals_df"]
    w_df = data["work_orders_df"]

    open_deals = d_df[d_df["isOpen"]]
    won_deals = d_df[d_df["isWon"]]
    total_pipeline = float(open_deals["Deal Value"].sum())
    won_val = float(won_deals["Deal Value"].sum())
    win_rate = round(len(won_deals) / len(d_df) * 100) if len(d_df) > 0 else 0

    contract_val = float(w_df["Amount Incl GST"].sum())
    billed_val = float(w_df["Billed Incl GST"].sum())
    collected_val = float(w_df["Collected Amount"].sum())
    receivable_val = float(w_df["Amount Receivable"].sum())
    stuck_wos = len(w_df[w_df["isStuck"]])
    active_wos = len(w_df[w_df["isOngoing"] | ~w_df["isCompleted"]])

    return {
        "pipeline": {
            "totalOpenDeals": len(open_deals),
            "totalPipelineValue": round(total_pipeline),
            "wonDealsCount": len(won_deals),
            "wonValue": round(won_val),
            "winRate": win_rate,
        },
        "operations": {
            "totalWorkOrders": len(w_df),
            "totalContractValue": round(contract_val),
            "totalBilled": round(billed_val),
            "totalCollected": round(collected_val),
            "totalReceivable": round(receivable_val),
            "ongoingWorkOrders": active_wos,
            "stuckWorkOrders": stuck_wos,
            "billingRate": round(billed_val / contract_val * 100) if contract_val > 0 else 0,
            "collectionRate": round(collected_val / billed_val * 100) if billed_val > 0 else 0,
        },
        "source": data["source"],
        "lastSync": data["last_sync"],
    }


def get_leadership_summary() -> Dict[str, Any]:
    """Prepare comprehensive executive leadership briefing."""
    kpis = get_kpis()
    data = get_data()
    d_df = data["deals_df"]
    w_df = data["work_orders_df"]

    # Top pipeline sector
    open_df = d_df[d_df["isOpen"]]
    top_sec_pipeline = open_df.groupby("Sector")["Deal Value"].sum().sort_values(ascending=False).head(1)
    top_pipeline_sec = {"sector": top_sec_pipeline.index[0], "value": round(float(top_sec_pipeline.iloc[0]))} if len(top_sec_pipeline) > 0 else None

    # Top receivable sector
    top_sec_recv = w_df.groupby("Sector")["Amount Receivable"].sum().sort_values(ascending=False).head(1)
    top_recv_sec = {"sector": top_sec_recv.index[0], "value": round(float(top_sec_recv.iloc[0]))} if len(top_sec_recv) > 0 else None

    # Key risks
    risks = []
    stuck_count = len(w_df[w_df["isStuck"]])
    if stuck_count > 0:
        risks.append(f"{stuck_count} work order projects are currently stalled/stuck.")
    
    high_recv = w_df[w_df["Amount Receivable"] > 1000000]
    if len(high_recv) > 0:
        risks.append(f"{len(high_recv)} client accounts have overdue balances exceeding ₹10 Lakhs.")

    # High value late-stage deals
    late_stage = (
        open_df[open_df["Stage Order"].between(5, 8)]
        .sort_values(by="Deal Value", ascending=False)
        .head(5)[["Deal Name", "Sector", "Deal Stage", "Deal Value"]]
        .to_dict(orient="records")
    )
    for d in late_stage:
        d["Deal Value"] = round(float(d["Deal Value"]))

    return {
        "pipeline": kpis["pipeline"],
        "operations": kpis["operations"],
        "topPipelineSector": top_pipeline_sec,
        "topReceivableSector": top_recv_sec,
        "lateStageOpportunities": late_stage,
        "risks": risks,
    }

File: server/data/analytics.js
/**
 * Analytics Engine — deterministic BI calculations.
 * All calculations happen here; the AI agent never touches raw data.
 */

const { getCache } = require('./cache');

function fmt(n) { return Math.round(n); }

// ══════════════════════════════════════════════════════════════════════════
// DEALS ANALYTICS
// ══════════════════════════════════════════════════════════════════════════

function queryDeals({ sector, dealStatus, dealStage, ownerCode, minValue, maxValue, groupBy, limit = 50 } = {}) {
  const { deals, dealWarnings } = getCache();
  let filtered = [...deals];

  // Apply filters
  if (sector && sector !== 'all') {
    const s = sector.toLowerCase();
    filtered = filtered.filter(d => d.sector.toLowerCase().includes(s));
  }
  if (dealStatus && dealStatus !== 'all') {
    filtered = filtered.filter(d => d.dealStatus.toLowerCase() === dealStatus.toLowerCase());
  }
  if (dealStage && dealStage !== 'all') {
    filtered = filtered.filter(d => d.dealStage.toLowerCase().includes(dealStage.toLowerCase()));
  }
  if (ownerCode && ownerCode !== 'all') {
    filtered = filtered.filter(d => d.ownerCode === ownerCode);
  }
  if (minValue) filtered = filtered.filter(d => d.dealValue >= minValue);
  if (maxValue) filtered = filtered.filter(d => d.dealValue <= maxValue);

  const totalValue = filtered.reduce((s, d) => s + d.dealValue, 0);
  const wonDeals = filtered.filter(d => d.isWon);
  const openDeals = filtered.filter(d => d.isOpen);
  const lostDeals = filtered.filter(d => d.isLost);
  const winRate = filtered.length ? Math.round((wonDeals.length / filtered.length) * 100) : 0;

  // Group by if requested
  let groups = null;
  if (groupBy) {
    const map = {};
    for (const d of filtered) {
      const key = d[groupBy] || 'Unknown';
      if (!map[key]) map[key] = { count: 0, totalValue: 0, wonCount: 0 };
      map[key].count++;
      map[key].totalValue += d.dealValue;
      if (d.isWon) map[key].wonCount++;
    }
    groups = Object.entries(map)
      .sort((a, b) => b[1].totalValue - a[1].totalValue)
      .map(([key, v]) => ({
        name: key,
        count: v.count,
        totalValue: fmt(v.totalValue),
        wonCount: v.wonCount,
        winRate: Math.round((v.wonCount / v.count) * 100),
      }));
  }

  // Stage breakdown
  const stageMap = {};
  for (const d of filtered) {
    const stage = d.dealStage || 'Unknown';
    if (!stageMap[stage]) stageMap[stage] = { count: 0, value: 0 };
    stageMap[stage].count++;
    stageMap[stage].value += d.dealValue;
  }
  const stageBreakdown = Object.entries(stageMap)
    .sort((a, b) => b[1].value - a[1].value)
    .map(([s, v]) => ({ stage: s, count: v.count, value: fmt(v.value) }));

  // Sector breakdown
  const sectorMap = {};
  for (const d of filtered) {
    const sec = d.sector || 'Unknown';
    if (!sectorMap[sec]) sectorMap[sec] = { count: 0, value: 0 };
    sectorMap[sec].count++;
    sectorMap[sec].value += d.dealValue;
  }
  const sectorBreakdown = Object.entries(sectorMap)
    .sort((a, b) => b[1].value - a[1].value)
    .map(([s, v]) => ({ sector: s, count: v.count, value: fmt(v.value) }));

  // Top deals
  const topDeals = [...filtered]
    .sort((a, b) => b.dealValue - a.dealValue)
    .slice(0, limit)
    .map(d => ({
      dealName: d.dealName,
      clientCode: d.clientCode,
      sector: d.sector,
      stage: d.dealStage,
      status: d.dealStatus,
      value: fmt(d.dealValue),
      probability: d.closureProbability,
      tentativeClose: d.tentativeCloseDate,
    }));

  return {
    summary: {
      totalDeals: filtered.length,
      totalValue: fmt(totalValue),
      wonDeals: wonDeals.length,
      wonValue: fmt(wonDeals.reduce((s, d) => s + d.dealValue, 0)),
      openDeals: openDeals.length,
      openValue: fmt(openDeals.reduce((s, d) => s + d.dealValue, 0)),
      lostDeals: lostDeals.length,
      winRate,
    },
    stageBreakdown,
    sectorBreakdown,
    groups,
    topDeals,
    dataWarnings: dealWarnings.slice(0, 5),
    filtersApplied: { sector, dealStatus, dealStage, ownerCode },
  };
}

// ══════════════════════════════════════════════════════════════════════════
// WORK ORDER ANALYTICS
// ══════════════════════════════════════════════════════════════════════════

function queryWorkOrders({ sector, executionStatus, invoiceStatus, billingStatus, groupBy, limit = 50 } = {}) {
  const { workOrders, workOrderWarnings } = getCache();
  let filtered = [...workOrders];

  if (sector && sector !== 'all') {
    const s = sector.toLowerCase();
    filtered = filtered.filter(w => w.sector.toLowerCase().includes(s));
  }
  if (executionStatus && executionStatus !== 'all') {
    filtered = filtered.filter(w => w.executionStatus.toLowerCase().includes(executionStatus.toLowerCase()));
  }
  if (invoiceStatus && invoiceStatus !== 'all') {
    filtered = filtered.filter(w => w.invoiceStatus.toLowerCase().includes(invoiceStatus.toLowerCase()));
  }
  if (billingStatus && billingStatus !== 'all') {
    filtered = filtered.filter(w => w.billingStatus.toLowerCase().includes(billingStatus.toLowerCase()));
  }

  const totalContract = filtered.reduce((s, w) => s + w.amountInclGST, 0);
  const totalBilled = filtered.reduce((s, w) => s + w.billedInclGST, 0);
  const totalCollected = filtered.reduce((s, w) => s + w.collectedAmount, 0);
  const totalReceivable = filtered.reduce((s, w) => s + w.receivable, 0);
  const totalUnbilled = filtered.reduce((s, w) => s + Math.max(0, w.amountInclGST - w.billedInclGST), 0);

  const completed = filtered.filter(w => w.isCompleted);
  const ongoing = filtered.filter(w => w.isOngoing);
  const stuck = filtered.filter(w => w.isStuck);
  const notStarted = filtered.filter(w => w.executionStatus === 'Not Started');

  // Sector breakdown
  const sectorMap = {};
  for (const w of filtered) {
    const sec = w.sector || 'Unknown';
    if (!sectorMap[sec]) sectorMap[sec] = { count: 0, contract: 0, billed: 0, collected: 0, receivable: 0 };
    sectorMap[sec].count++;
    sectorMap[sec].contract += w.amountInclGST;
    sectorMap[sec].billed += w.billedInclGST;
    sectorMap[sec].collected += w.collectedAmount;
    sectorMap[sec].receivable += w.receivable;
  }
  const sectorBreakdown = Object.entries(sectorMap)
    .sort((a, b) => b[1].contract - a[1].contract)
    .map(([sec, v]) => ({
      sector: sec,
      count: v.count,
      contractValue: fmt(v.contract),
      billedValue: fmt(v.billed),
      collectedValue: fmt(v.collected),
      receivableValue: fmt(v.receivable),
      billingRate: v.contract ? Math.round((v.billed / v.contract) * 100) : 0,
      collectionRate: v.billed ? Math.round((v.collected / v.billed) * 100) : 0,
    }));

  // Status breakdown
  const statusMap = {};
  for (const w of filtered) {
    const s = w.executionStatus || 'Unknown';
    if (!statusMap[s]) statusMap[s] = { count: 0, value: 0 };
    statusMap[s].count++;
    statusMap[s].value += w.amountInclGST;
  }
  const statusBreakdown = Object.entries(statusMap)
    .sort((a, b) => b[1].count - a[1].count)
    .map(([s, v]) => ({ status: s, count: v.count, value: fmt(v.value) }));

  // Group by if requested
  let groups = null;
  if (groupBy) {
    const map = {};
    for (const w of filtered) {
      const key = w[groupBy] || 'Unknown';
      if (!map[key]) map[key] = { count: 0, contract: 0, billed: 0, collected: 0, receivable: 0 };
      map[key].count++;
      map[key].contract += w.amountInclGST;
      map[key].billed += w.billedInclGST;
      map[key].collected += w.collectedAmount;
      map[key].receivable += w.receivable;
    }
    groups = Object.entries(map)
      .sort((a, b) => b[1].contract - a[1].contract)
      .map(([key, v]) => ({
        name: key,
        count: v.count,
        contractValue: fmt(v.contract),
        billedValue: fmt(v.billed),
        collectedValue: fmt(v.collected),
        receivableValue: fmt(v.receivable),
      }));
  }

  // Top work orders by receivable
  const topWOs = [...filtered]
    .filter(w => w.receivable > 0)
    .sort((a, b) => b.receivable - a.receivable)
    .slice(0, limit)
    .map(w => ({
      dealName: w.dealName,
      customerCode: w.customerCode,
      sector: w.sector,
      executionStatus: w.executionStatus,
      invoiceStatus: w.invoiceStatus,
      contractValue: fmt(w.amountInclGST),
      billedValue: fmt(w.billedInclGST),
      collectedValue: fmt(w.collectedAmount),
      receivable: fmt(w.receivable),
    }));

  return {
    summary: {
      totalWorkOrders: filtered.length,
      totalContractValue: fmt(totalContract),
      totalBilled: fmt(totalBilled),
      totalCollected: fmt(totalCollected),
      totalReceivable: fmt(totalReceivable),
      totalUnbilled: fmt(totalUnbilled),
      completedCount: completed.length,
      ongoingCount: ongoing.length,
      stuckCount: stuck.length,
      notStartedCount: notStarted.length,
      billingRate: totalContract ? Math.round((totalBilled / totalContract) * 100) : 0,
      collectionRate: totalBilled ? Math.round((totalCollected / totalBilled) * 100) : 0,
    },
    sectorBreakdown,
    statusBreakdown,
    groups,
    topReceivables: topWOs,
    dataWarnings: workOrderWarnings.slice(0, 5),
    filtersApplied: { sector, executionStatus, invoiceStatus, billingStatus },
  };
}

// ══════════════════════════════════════════════════════════════════════════
// CROSS-BOARD SUMMARY (best-effort join on sector)
// ══════════════════════════════════════════════════════════════════════════

function crossBoardSummary({ sector } = {}) {
  const deals = queryDeals({ sector });
  const wos = queryWorkOrders({ sector });

  return {
    sector: sector || 'All Sectors',
    pipeline: deals.summary,
    operations: wos.summary,
    sectorPipelineBreakdown: deals.sectorBreakdown,
    sectorOperationsBreakdown: wos.sectorBreakdown,
    note: 'Cross-board join is by sector (best-effort). No shared unique key exists between boards.',
    dataWarnings: [...deals.dataWarnings, ...wos.dataWarnings],
  };
}

// ══════════════════════════════════════════════════════════════════════════
// TOP-LEVEL KPIs
// ══════════════════════════════════════════════════════════════════════════

function getKPIs() {
  const { deals, workOrders } = getCache();

  const openDeals = deals.filter(d => d.isOpen);
  const wonDeals = deals.filter(d => d.isWon);
  const totalPipeline = openDeals.reduce((s, d) => s + d.dealValue, 0);
  const wonValue = wonDeals.reduce((s, d) => s + d.dealValue, 0);
  const winRate = deals.length ? Math.round((wonDeals.length / deals.length) * 100) : 0;

  const totalContract = workOrders.reduce((s, w) => s + w.amountInclGST, 0);
  const totalReceivable = workOrders.reduce((s, w) => s + w.receivable, 0);
  const totalBilled = workOrders.reduce((s, w) => s + w.billedInclGST, 0);
  const totalCollected = workOrders.reduce((s, w) => s + w.collectedAmount, 0);
  const ongoingWOs = workOrders.filter(w => w.isOngoing || !w.isCompleted).length;
  const stuckWOs = workOrders.filter(w => w.isStuck).length;

  return {
    pipeline: {
      totalOpenDeals: openDeals.length,
      totalPipelineValue: fmt(totalPipeline),
      winRate,
      wonDealsCount: wonDeals.length,
      wonValue: fmt(wonValue),
    },
    operations: {
      totalWorkOrders: workOrders.length,
      totalContractValue: fmt(totalContract),
      totalBilled: fmt(totalBilled),
      totalCollected: fmt(totalCollected),
      totalReceivable: fmt(totalReceivable),
      ongoingWorkOrders: ongoingWOs,
      stuckWorkOrders: stuckWOs,
      billingRate: totalContract ? Math.round((totalBilled / totalContract) * 100) : 0,
      collectionRate: totalBilled ? Math.round((totalCollected / totalBilled) * 100) : 0,
    },
    lastSync: getCache().lastSync,
  };
}

// ══════════════════════════════════════════════════════════════════════════
// LEADERSHIP SUMMARY
// ══════════════════════════════════════════════════════════════════════════

function getLeadershipSummary() {
  const kpis = getKPIs();
  const { deals, workOrders } = getCache();

  // Top sector by pipeline
  const sectorPipeline = {};
  for (const d of deals.filter(d => d.isOpen)) {
    const s = d.sector || 'Unknown';
    sectorPipeline[s] = (sectorPipeline[s] || 0) + d.dealValue;
  }
  const topPipelineSector = Object.entries(sectorPipeline).sort((a, b) => b[1] - a[1])[0];

  // Top sector by receivables
  const sectorReceivable = {};
  for (const w of workOrders) {
    const s = w.sector || 'Unknown';
    sectorReceivable[s] = (sectorReceivable[s] || 0) + w.receivable;
  }
  const topReceivableSector = Object.entries(sectorReceivable).sort((a, b) => b[1] - a[1])[0];

  // Risks
  const risks = [];
  const stuckWOs = workOrders.filter(w => w.isStuck);
  const highValueDeals = deals.filter(d => d.isOpen && d.dealValue > 5000000);
  const overdueReceivables = workOrders.filter(w => w.receivable > 1000000);

  if (stuckWOs.length > 0) risks.push(`${stuckWOs.length} work order(s) are stuck/paused`);
  if (highValueDeals.length > 0) risks.push(`${highValueDeals.length} high-value deals (>₹50L) still in pipeline`);
  if (overdueReceivables.length > 0) risks.push(`${overdueReceivables.length} work orders have receivables >₹10L`);

  // Top late-stage deals
  const lateStageDeals = deals
    .filter(d => d.isOpen && d.stageOrder >= 5 && d.stageOrder <= 8)
    .sort((a, b) => b.dealValue - a.dealValue)
    .slice(0, 5)
    .map(d => ({ name: d.dealName, sector: d.sector, value: fmt(d.dealValue), stage: d.dealStage }));

  return {
    pipeline: kpis.pipeline,
    operations: kpis.operations,
    topPipelineSector: topPipelineSector ? { sector: topPipelineSector[0], value: fmt(topPipelineSector[1]) } : null,
    topReceivableSector: topReceivableSector ? { sector: topReceivableSector[0], value: fmt(topReceivableSector[1]) } : null,
    lateStageDeals,
    risks,
    generatedAt: new Date().toISOString(),
  };
}

module.exports = { queryDeals, queryWorkOrders, crossBoardSummary, getKPIs, getLeadershipSummary };
